### PR TITLE
Restrict `NameSegment` operations

### PR DESCRIPTION
**Pull request by sellout**
## Overview

With `OverloadedStrings` enabled globally and an `IsString` instance, the `newtype` was rendered useless.

## Implementation notes

This removes the `IsString` instance, extracts the `NameSegment` constructor/eliminator to a `.Internal` module and has `Unison.NameSegment` only re-export the type and sentinel segment names.

This forces cascading changes, including eliminating a bunch of magic literals scattered throughout the code.

### CODE REVIEW

**Comment by aryairani** [comment]
@mitchellwrosen Could you review this one too

File: codebase2/core/Unison/NameSegment.hs [hunk 1]
```diff
@@ -1,57 +1,15 @@
...
-
-defaultPatchSegment :: NameSegment
-defaultPatchSegment =
-  "patch"
-
-docSegment :: NameSegment
-docSegment =
-  "doc"
-
+--  __TODO__: This should live in "Unison.Syntax.NameSegment", but it’s currently used in unison-core.
```
Comment (mitchellwrosen) [inline codebase2/core/Unison/NameSegment.hs]: As discussed, we are going to pull docSegment and friends back in here

File: unison-cli/src/Unison/Codebase/Editor/UriParser.hs [hunk 2]
```diff
@@ -62,9 +62,12 @@ type P = P.Parsec Void Text.Text
 readRemoteNamespaceParser :: ProjectBranchSpecifier branch -> P (ReadRemoteNamespace (These ProjectName branch))
 readRemoteNamespaceParser specifier =
   P.label "generic repo" $
-    ReadRemoteNamespaceGit <$> readGitRemoteNamespace
-      <|> ReadShare'ProjectBranch <$> projectAndBranchNamesParserInTheContextOfAlsoParsingLooseCodePaths specifier
-      <|> ReadShare'LooseCode <$> readShareLooseCode
+    ReadRemoteNamespaceGit
```
Comment (mitchellwrosen) [inline unison-cli/src/Unison/Codebase/Editor/UriParser.hs]: ormolu seems to have gotten the precedence wrong here and below?

**Review by mitchellwrosen — APPROVED** [review]

**Comment by aryairani** [comment]
Will y'all signal when the review comments are addressed to your respective satisfaction?

**Comment by mitchellwrosen** [comment]
@aryairani yeah – we reviewed over the phone, it is good to go aside from having conflicts and some ormolu things to work out

**Comment by aryairani** [comment]
Ok — go ahead and merge then when ready

**Comment by sellout** [comment]
Ok, should be all good. And merged in trunk.

**Comment by mitchellwrosen** [comment]
Looks like some tests have compiler errors @sellout

**Comment by sellout** [comment]
> Looks like some tests have compiler errors @sellout

Sorry – during merge conflict resolution I even said to myself “I’m going to let the compiler help me get this one right“, then forgot to build the tests before pushing. It should be fixed now (since I actually built & ran the tests).

**Pull request merged**